docs/haskell-use-cases.rst

```diff
@@ -308,20 +308,195 @@ However, you seldom want to use them directly. Cabal packages
+individual Cabal packages.
+
+This is how you import the Stackage LTS 14.0 snapshot ::
+
+  stack_snapshot(
+      name = "stackage",
+      snapshot = "lts-14.0",
+      packages = [
+          "base",
+          "optparse-applicative",
+      ],
+  )
+
+This will generate the labels ``@stackage//:base``, and
+``@stackage//:optparse-applicative``, which you can use in the ``deps``
+attribute of your Haskell targets. Note that ``base`` is a core package and its
+version is determined by the GHC toolchain and not the Stackage snapshot.
+
+Use the ``local_snapshot`` attribute to refer to a `custom Stack snapshot`_.
+
+Pinning
+^^^^^^^
+
+The ``stack_snapshot`` rule invokes ``stack`` for version and dependency
+resolution. By default this will happen on every fetch of the repository. You
+can enable pinning to avoid repeated fetching and to ensure that dependencies
+don't change, e.g. due to new Cabal revisions.
+
+1. Generate a lock file by running ``bazel run @stackage-unpinned//:pin``.
+2. Set the ``stack_snapshot_json`` attribute. ::
+
+     stack_snapshot(
+         ...
+         stack_snapshot_json = "//:stackage_snapshot.json",
+     )
+
+Repeat step 1 when you change the ``stack_snapshot`` definition, e.g. the
+Stackage snapshot or the list of packages.
+
+Version overrides or Hackage dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also depend on Hackage packages that are not part of a Stackage
+snapshot, or override the version of a package, by specifying the version in
+the ``packages`` attribute. ::
+
+  stack_snapshot(
+      ...
+      packages = [
+          ...
+          "optparse-helper-0.2.1.1",
+      ],
+  )
+
+C-library dependencies
+^^^^^^^^^^^^^^^^^^^^^^
+
+Some Hackage packages depend on C libraries. Bazel builds should be hermetic,
+therefore, such library dependencies should be managed by Bazel and declared
+explicitly. ::
+
+  stack_snapshot(
+      ...
+      packages = [
+          ...
+          "zlib",
+      ],
+      extra_deps = {
+          "zlib": ["@zlib-deps//:libz"],
+      },
+  )
+
+This declares that the Stackage package ``zlib`` has an additional dependency
+``@zlib-deps//:libz``. The C library ``libz`` could be imported using
+rules_nixpkgs, or fetched and built by Bazel as follows. ::
+
+  http_archive(
+      name = "zlib-deps",
+      build_file_content = """
+  load("@rules_cc//cc:defs.bzl", "cc_library")
+  cc_library(
+      name = "libz",
+      # The indirection enforces the library name `libz.so`,
+      # otherwise Cabal won't find it.
+      srcs = [":z"],
+      hdrs = glob(["*.h"]),
+      includes = ["."],
+      visibility = ["//visibility:public"],
+  )
+  cc_library(name = "z", srcs = glob(["*.c"]), hdrs = glob(["*.h"]))
+  """,
+      sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1",
+      strip_prefix = "zlib-1.2.11",
+      urls = ["http://zlib.net/zlib-1.2.11.tar.gz"],
+  )
+
+Vendoring packages
```

Comment:
The difference between `vendored_packages` and `extra_deps` is unclear to me.

Reply:
The two attributes don't have the same type. `extra_deps` serves to state additional dependencies to Haskell packages. `vendored_packages` serves to replace Haskell packages with targets defined elsewhere. According to the respective sections. What could be said to make the distinction clearer?

Reply:
Haa, got it. Perhaps not using the same `zlib` (the C library and the Haskell package) in the example may be more clear?

Reply:
I've changed it to use a vendored version of `hashable` and to illustrate on `unordered-containers` that the vendored version is injected into the dependency graph.